File: web/src/types/notification.ts
// Notification types - mirrors internal/domain/notification.go

export type NotificationChannel = 'email' | 'sms' | 'whatsapp' | 'internal'
export type NotificationStatus = 'pending' | 'sent' | 'delivered' | 'failed' | 'cancelled'
export type NotificationType = 'payment_reminder' | 'overdue_notice' | 'loan_expiry' | 'promotional' | 'system'

export const NOTIFICATION_CHANNELS: { value: NotificationChannel; label: string }[] = [
  { value: 'email', label: 'Email' },
  { value: 'sms', label: 'SMS' },
  { value: 'whatsapp', label: 'WhatsApp' },
  { value: 'internal', label: 'Interna' },
]

export const NOTIFICATION_STATUSES: { value: NotificationStatus; label: string; color: string }[] = [
  { value: 'pending', label: 'Pendiente', color: 'yellow' },
  { value: 'sent', label: 'Enviada', color: 'blue' },
  { value: 'delivered', label: 'Entregada', color: 'green' },
  { value: 'failed', label: 'Fallida', color: 'red' },
  { value: 'cancelled', label: 'Cancelada', color: 'gray' },
]

export interface NotificationTemplate {
  id: number
  name: string
  code: string
  channel: NotificationChannel
  notification_type: NotificationType
  subject?: string
  content: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface Notification {
  id: number
  customer_id: number
  branch_id?: number
  template_id?: number
  channel: NotificationChannel
  notification_type: NotificationType
  subject?: string
  content: string
  status: NotificationStatus
  scheduled_at?: string
  sent_at?: string
  delivered_at?: string
  failed_at?: string
  error_message?: string
  reference_type?: string
  reference_id?: number
  created_by?: number
  created_at: string
  updated_at: string
}

export interface InternalNotification {
  id: number
  user_id: number
  branch_id?: number
  title: string
  message: string
  notification_type: string
  reference_type?: string
  reference_id?: number
  is_read: boolean
  read_at?: string
  created_by?: number
  created_at: string
}

export interface CustomerNotificationPreference {
  id: number
  customer_id: number
  email_enabled: boolean
  sms_enabled: boolean
  whatsapp_enabled: boolean
  payment_reminders: boolean
  overdue_notices: boolean
  promotional: boolean
  created_at: string
  updated_at: string
}

export interface CreateNotificationTemplateInput {
  name: string
  code: string
  channel: NotificationChannel
  notification_type: NotificationType
  subject?: string
  content: string
}

export interface UpdateNotificationTemplateInput {
  name?: string
  subject?: string
  content?: string
  is_active?: boolean
}

export interface CreateNotificationInput {
  customer_id: number
  channel: NotificationChannel
  notification_type: NotificationType
  subject?: string
  content: string
  scheduled_at?: string
  reference_type?: string
  reference_id?: number
}

export interface CreateFromTemplateInput {
  customer_id: number
  template_code: string
  variables?: Record<string, string>
  scheduled_at?: string
  reference_type?: string
  reference_id?: number
}

export interface CreateInternalNotificationInput {
  user_id: number
  title: string
  message: string
  notification_type?: string
  reference_type?: string
  reference_id?: number
}

export interface UpdateNotificationPreferencesInput {
  email_enabled?: boolean
  sms_enabled?: boolean
  whatsapp_enabled?: boolean
  payment_reminders?: boolean
  overdue_notices?: boolean
  promotional?: boolean
}

export interface NotificationListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  customer_id?: number
  branch_id?: number
  channel?: NotificationChannel
  status?: NotificationStatus
  notification_type?: NotificationType
  date_from?: string
  date_to?: string
}

File: internal/domain/notification.go
package domain

import "time"

// Notification types
const (
	NotificationTypeLoanDueReminder   = "loan_due_reminder"
	NotificationTypeLoanOverdue       = "loan_overdue"
	NotificationTypeMinimumPaymentDue = "minimum_payment_due"
	NotificationTypePaymentReceived   = "payment_received"
	NotificationTypeLoanConfiscated   = "loan_confiscated"
	NotificationTypeItemForSale       = "item_for_sale"
	NotificationTypeItemSold          = "item_sold"
	NotificationTypePromotion         = "promotion"
	NotificationTypeLoyaltyPoints     = "loyalty_points"
	NotificationTypeGeneral           = "general"
)

// Notification channels
const (
	NotificationChannelEmail    = "email"
	NotificationChannelSMS      = "sms"
	NotificationChannelWhatsApp = "whatsapp"
	NotificationChannelPush     = "push"
	NotificationChannelInternal = "internal"
)

// Notification status
const (
	NotificationStatusPending   = "pending"
	NotificationStatusSent      = "sent"
	NotificationStatusDelivered = "delivered"
	NotificationStatusFailed    = "failed"
	NotificationStatusCancelled = "cancelled"
)

// NotificationTemplate represents a template for notifications
type NotificationTemplate struct {
	ID              int64  `json:"id"`
	NotificationType string `json:"notification_type"`
	Channel         string `json:"channel"`
	Name            string `json:"name"`
	Subject         string `json:"subject,omitempty"`
	BodyTemplate    string `json:"body_template"`
	IsActive        bool   `json:"is_active"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// Notification represents a notification to be sent to a customer
type Notification struct {
	ID               int64     `json:"id"`
	CustomerID       int64     `json:"customer_id"`
	Customer         *Customer `json:"customer,omitempty"`
	BranchID         *int64    `json:"branch_id,omitempty"`
	Branch           *Branch   `json:"branch,omitempty"`
	NotificationType string    `json:"notification_type"`
	Channel          string    `json:"channel"`

	// Content
	Subject string `json:"subject,omitempty"`
	Body    string `json:"body"`

	// Reference
	ReferenceType string `json:"reference_type,omitempty"`
	ReferenceID   *int64 `json:"reference_id,omitempty"`

	// Delivery
	Status       string     `json:"status"`
	ScheduledFor *time.Time `json:"scheduled_for,omitempty"`
	SentAt       *time.Time `json:"sent_at,omitempty"`
	DeliveredAt  *time.Time `json:"delivered_at,omitempty"`
	FailedAt     *time.Time `json:"failed_at,omitempty"`
	FailureReason string    `json:"failure_reason,omitempty"`
	RetryCount   int        `json:"retry_count"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// IsPending checks if notification is pending
func (n *Notification) IsPending() bool {
	return n.Status == NotificationStatusPending
}

// IsSent checks if notification was sent
func (n *Notification) IsSent() bool {
	return n.Status == NotificationStatusSent
}

// IsDelivered checks if notification was delivered
func (n *Notification) IsDelivered() bool {
	return n.Status == NotificationStatusDelivered
}

// IsFailed checks if notification failed
func (n *Notification) IsFailed() bool {
	return n.Status == NotificationStatusFailed
}

// CanRetry checks if notification can be retried
func (n *Notification) CanRetry() bool {
	return n.Status == NotificationStatusFailed && n.RetryCount < 3
}

// CustomerNotificationPreference represents customer preferences for notifications
type CustomerNotificationPreference struct {
	ID               int64  `json:"id"`
	CustomerID       int64  `json:"customer_id"`
	NotificationType string `json:"notification_type"`
	Channel          string `json:"channel"`
	IsEnabled        bool   `json:"is_enabled"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// InternalNotification represents a notification for internal users
type InternalNotification struct {
	ID       int64  `json:"id"`
	UserID   int64  `json:"user_id"`
	User     *User  `json:"user,omitempty"`
	BranchID *int64 `json:"branch_id,omitempty"`
	Branch   *Branch `json:"branch,omitempty"`

	// Content
	Title   string `json:"title"`
	Message string `json:"message"`
	Type    string `json:"type"` // info, warning, error, success

	// Reference
	ReferenceType string `json:"reference_type,omitempty"`
	ReferenceID   *int64 `json:"reference_id,omitempty"`
	ActionURL     string `json:"action_url,omitempty"`

	// Status
	IsRead   bool       `json:"is_read"`
	ReadAt   *time.Time `json:"read_at,omitempty"`

	CreatedAt time.Time `json:"created_at"`
}

// MarkAsRead marks the notification as read
func (n *InternalNotification) MarkAsRead() {
	n.IsRead = true
	now := time.Now()
	n.ReadAt = &now
}
